File: scripts/qa/validate_surge_rulesets.py
#!/usr/bin/env python3
"""Validate Surge .list rulesets for known Surge-invalid patterns."""

from __future__ import annotations

import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "scripts"))

from hub.surge_compliance import validate_surge_ruleset_line  # noqa: E402

SURGE_DIR = ROOT / "rulesets" / "surge-shadowrocket"
ADBLOCK_DIR = ROOT / "rulesets" / "AdBlock"


def validate_directory(directory: Path, *, allow_domain_regex: bool) -> list[str]:
    all_issues: list[str] = []
    if not directory.is_dir():
        return [f"missing directory: {directory}"]
    for path in sorted(directory.glob("*.list")):
        if "skk_upstream" in path.parts:
            continue
        text = path.read_text(encoding="utf-8", errors="ignore")
        for i, line in enumerate(text.splitlines(), 1):
            err = validate_surge_ruleset_line(line, allow_domain_regex=allow_domain_regex)
            if err:
                all_issues.append(f"{path.name}:{i}: {err}")
    return all_issues


def main() -> int:
    issues: list[str] = []
    if SURGE_DIR.is_dir():
        issues.extend(validate_directory(SURGE_DIR, allow_domain_regex=False))
    if ADBLOCK_DIR.is_dir():
        issues.extend(validate_directory(ADBLOCK_DIR, allow_domain_regex=True))
    if issues:
        print("Surge ruleset validation failed:", file=sys.stderr)
        for item in issues[:50]:
            print(f"  - {item}", file=sys.stderr)
        if len(issues) > 50:
            print(f"  ... and {len(issues) - 50} more", file=sys.stderr)
        return 1
    print("Surge ruleset validation OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: create/qa/validate_surge_rulesets.go
package qa

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/nowaytouse/script_hub/create/hub"
)

func validateDirectory(directory string, allowDomainRegex bool) []string {
	var allIssues []string
	if !hub.ValidateDirExists(directory, "") {
		return []string{fmt.Sprintf("missing directory: %s", directory)}
	}

	filepath.WalkDir(directory, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		if !strings.HasSuffix(d.Name(), ".list") {
			return nil
		}
		if strings.Contains(filepath.ToSlash(path), "skk_upstream") {
			return nil
		}

		text := hub.ReadFileString(path)
		lines := strings.Split(text, "\n")
		for i, line := range lines {
			errStr := hub.ValidateSurgeRulesetLine(line, allowDomainRegex)
			if errStr != nil {
				allIssues = append(allIssues, fmt.Sprintf("%s:%d: %s", d.Name(), i+1, *errStr))
			}
		}
		return nil
	})

	return allIssues
}

func RunValidateSurgeRulesets() int {
	surgeDir := filepath.Join(hub.ROOT, "rulesets/list")
	adblockDir := filepath.Join(hub.ROOT, "rulesets/AdBlock")

	var issues []string

	if info, err := os.Stat(surgeDir); err == nil && info.IsDir() {
		issues = append(issues, validateDirectory(surgeDir, false)...)
	}
	if info, err := os.Stat(adblockDir); err == nil && info.IsDir() {
		issues = append(issues, validateDirectory(adblockDir, true)...)
	}

	if len(issues) > 0 {
		fmt.Fprintln(os.Stderr, "Surge ruleset validation failed:")
		for i, item := range issues {
			if i >= 50 {
				break
			}
			fmt.Fprintf(os.Stderr, "  - %s\n", item)
		}
		if len(issues) > 50 {
			fmt.Fprintf(os.Stderr, "  ... and %d more\n", len(issues)-50)
		}
		return 1
	}

	fmt.Println("Surge ruleset validation OK")
	return 0
}
